mongomock/collection.py

```diff
@@ -2079,6 +2080,13 @@ def _set_updater(doc, field_name, value):
             # bson validation
             BSON.encode({field_name: value}, check_keys=True)
         doc[field_name] = value
+    if isinstance(doc, list):
+        try:
+            doc[int(field_name)] = value
```

Comment:
What if field index is negative? I think that it's not correct for the Mongo syntax to have a field `foo.-3.bar` so you don't need to handle it correctly, however mongomock should raise an error instead of handling it silently.